common/src/time_getter.rs

```diff
@@ -13,33 +13,137 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::{
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+    time::{Duration, Instant},
+};
+
 use crate::primitives::time;
-use std::{sync::Arc, time::Duration};
 
-pub type TimeGetterFn = dyn Fn() -> Duration + Send + Sync;
+pub trait TimeGetterFn: Send + Sync {
+    fn system_time(&self) -> Duration;
+
+    fn instant(&self) -> Instant;
+}
 
 /// A function wrapper that contains the function that will be used to get the current time in chainstate
 #[derive(Clone)]
 pub struct TimeGetter {
-    f: Arc<TimeGetterFn>,
+    f: Arc<dyn TimeGetterFn>,
 }
 
 impl TimeGetter {
-    pub fn new(f: Arc<TimeGetterFn>) -> Self {
+    pub fn new(f: Arc<dyn TimeGetterFn>) -> Self {
         Self { f }
     }
 
+    // NOTE: `get_instant` won't work correctly if the counter goes backwards
+    pub fn mocked_seconds(seconds: Arc<AtomicU64>) -> Self {
+        Self::new(Arc::new(MockedMsecTimeGetterFn::new(seconds, 1000)))
+    }
+
+    // NOTE: `get_instant` won't work correctly if the counter goes backwards
+    pub fn mocked_milliseconds(milliseconds: Arc<AtomicU64>) -> Self {
+        Self::new(Arc::new(MockedMsecTimeGetterFn::new(milliseconds, 1)))
+    }
+
     pub fn get_time(&self) -> Duration {
-        (self.f)()
+        self.f.system_time()
+    }
+
+    pub fn get_instant(&self) -> Instant {
+        self.f.instant()
     }
 
-    pub fn getter(&self) -> &TimeGetterFn {
+    pub fn getter(&self) -> &dyn TimeGetterFn {
         &*self.f
     }
 }
 
 impl Default for TimeGetter {
     fn default() -> Self {
-        Self::new(Arc::new(time::get))
+        Self::new(Arc::new(DefaultTimeGetterFn::new()))
+    }
+}
+
+pub struct DefaultTimeGetterFn;
+
+impl DefaultTimeGetterFn {
+    pub fn new() -> Self {
+        Self
+    }
+}
+
+impl TimeGetterFn for DefaultTimeGetterFn {
+    fn system_time(&self) -> Duration {
+        time::get_system_time()
+    }
+
+    fn instant(&self) -> Instant {
+        time::get_instant_time()
+    }
+}
+
+pub struct MockedMsecTimeGetterFn {
```

Comment:
I think this can go.

Reply:
I think you missed that `MockedMsecTimeGetterFn` is used in unit tests (from the `TimeGetter::mocked_seconds` constructor). I don't see how we can remove it.

Reply:
I don't think this should exist. We should just use the absolute time getters and instant getters.

Reply:
Can you please elaborate on this? I'm still not sure what should be done here.
BTW, I made `MockedMsecTimeGetterFn` and `DefaultTimeGetterFn` private so they are not even visible from the outside.

Reply:
Like I mentioned in the meeting, there has to be exactly one source of time in TimeGetter. There shouldn't be the possibility to query different kinds of time.

Reply:
Maybe it would be better to close this PR and keep using `tokio::time::Instant` in p2p (where absolute time is not needed)?

Reply:
The current usage of instant is a problem because it's decoupled from mocked time. We need this PR.